File: leetcode/509.java
class Solution {
    public int fib(int n) {
        return fib2(n, 0, 1);
    }
    
    public static int fib2(int n, int a, int b) {
        if (n == 0) {
            return a;
        } else if (n == 1) {
            return b;
        } else {
            return fib2(n-1, b, a+b);
        }
    }    
}

File: leetcode/509.scala
import scala.annotation.tailrec

object Solution {
    def fib(n: Int): Int = {
        @tailrec
        def fib2(n: Int, a: Int = 0, b: Int = 1): Int = {
            n match {
                case 0 => a
                case 1 => b
                case _ => fib2(n-1, b, a+b)
            }
        }
        
        fib2(n)        
    }
}
